specutils/fitting/fitmodels.py

```diff
@@ -0,0 +1,507 @@
+
+        dispersion = dispersion[indices]
+        flux = flux[indices]
+
+    # In this case the window is the start and end points of where we should fit
+    elif window is not None and isinstance(window, tuple):
+        indices = np.nonzero((dispersion >= window[0]) & (dispersion < window[1]))
+
+        dispersion = dispersion[indices]
+        flux = flux[indices]
+
+    #
+    # Compound models with units can not be fit.
+    #
+    # Convert the model initial guess to the spectral
+    # units and then remove the units
+    #
+
+    model_unitless, dispersion_unitless, flux_unitless = _strip_units_from_model(model, spectrum)
+
+    #
+    # Do the fitting of spectrum to the model.
+    #
+
+    fit_model_unitless = fitter(model_unitless, dispersion_unitless, flux_unitless)
+
+    #
+    # Now add the units back onto the model....
+    #
+
+    fit_model = _add_units_to_model(fit_model_unitless, model, spectrum)
+
+    return fit_model
+
+
+def _strip_units_from_model(model_in, spectrum):
+    """
+    This method strips the units from the model, so the result can
+    be passed to the fitting routine. This is necessary as CoumpoundModel
+    with units does not work in the fitters.
+
+    Note:  When CompoundModel with units works in the fitters this method
+           can be removed.
+
+    Note: This assumes there are two types of models, those that are
+          based on `~astropy.modeling.models.PolynomialModel` and therefore
+          require the ``degree`` parameter when instantiating the class, and
+          "everything else" that does not require an "extra" parameter for
+          class instantiation.
+    """
+
+    #
+    # Get the dispersion and flux information from the spectrum
+    #
+
+    dispersion = spectrum.spectral_axis
+    dispersion_unit = spectrum.spectral_axis.unit
+
+    flux = spectrum.flux
+    flux_unit = spectrum.flux.unit
+
+    #
+    # Determine if a compound model
+    #
+
+    compound_model = model_in.n_submodels() > 1
+
+    if not compound_model:
+        # For this we are going to just make it a list so that we
+        # can use the looping structure below.
+        model_in = [model_in]
+    else:
+        # If it is a compound model then we are going to create the RPN
+        # representation of it which is a list that contains either astropy
+        # models or string representations of operators (e.g., '+' or '*').
+        model_in = [c.value for c in model_in._tree.traverse_postorder()]
+
+    #
+    # Run through each model in the list or compound model
+    #
+
+    model_out_stack = []
+    for sub_model in model_in:
+
+        #
+        # If it is an operator put onto the stack and move on...
+        #
+
+        if not isinstance(sub_model, Model):
+            model_out_stack.append(sub_model)
+            continue
+
+        #
+        # Make a new instance of the class.
+        #
+
+        if isinstance(sub_model, models.PolynomialModel):
```

Comment:
Alternative solution to this that's a bit more general:  
```
if any([p.default == p.empty for p in inspect.signature(sub_model.__class__).parameters.values() if p.kind==p.POSITIONAL_OR_KEYWORD])
```
You might want to unpack that into multiple lines (😉 ), but it looks to me like it does the job: if a model class has a keyword *without* a default, it will be treated like Polynomial.  I think that's basically the key thing here, right?  That there's no default?

You'd probably then have to programatically pick out the parameter name using the signature to replace the ``sub_model.degree`` in the next line, but that should work.

Reply:
(Also falls under "you can assign this to @eteq since it's his crazy idea, though :wink:)

Reply:
Ok. I would like to suggest this goes into a new PR once this PR is merged.